File: android-app/core/src/main/kotlin/com/bhan796/anagram/core/engine/PickerConstraints.kt
package com.bhan796.anagram.core.engine

import com.bhan796.anagram.core.model.LetterKind

class PickerConstraints(private val targetSlots: Int = 9) {
    fun allowedKinds(currentPicks: List<Char>, vowelCount: Int, consonantCount: Int): Set<LetterKind> {
        val remaining = targetSlots - currentPicks.size
        if (remaining <= 0) return emptySet()

        val allowed = mutableSetOf<LetterKind>()

        if (isPickAllowed(LetterKind.VOWEL, vowelCount, consonantCount, remaining)) {
            allowed += LetterKind.VOWEL
        }

        if (isPickAllowed(LetterKind.CONSONANT, vowelCount, consonantCount, remaining)) {
            allowed += LetterKind.CONSONANT
        }

        return allowed
    }

    fun isPickAllowed(
        kind: LetterKind,
        currentVowels: Int,
        currentConsonants: Int,
        remainingSlots: Int
    ): Boolean {
        if (remainingSlots <= 0) return false

        val nextVowels = currentVowels + if (kind == LetterKind.VOWEL) 1 else 0
        val nextConsonants = currentConsonants + if (kind == LetterKind.CONSONANT) 1 else 0

        return canStillSatisfyMinimums(nextVowels, nextConsonants, remainingSlots - 1)
    }

    private fun canStillSatisfyMinimums(vowels: Int, consonants: Int, remainingSlots: Int): Boolean {
        val neededVowels = maxOf(0, 1 - vowels)
        val neededConsonants = maxOf(0, 1 - consonants)
        return neededVowels + neededConsonants <= remainingSlots
    }
}

File: ios-app/AnagramArena/Sources/Core/Engine/PickerConstraints.swift
import Foundation

struct PickerConstraints {
    let targetSlots: Int

    init(targetSlots: Int = 9) {
        self.targetSlots = targetSlots
    }

    func allowedKinds(currentPicks: [Character], vowelCount: Int, consonantCount: Int) -> Set<LetterKind> {
        let remaining = targetSlots - currentPicks.count
        guard remaining > 0 else { return [] }

        var allowed: Set<LetterKind> = []

        if isPickAllowed(.vowel, currentVowels: vowelCount, currentConsonants: consonantCount, remainingSlots: remaining) {
            allowed.insert(.vowel)
        }

        if isPickAllowed(.consonant, currentVowels: vowelCount, currentConsonants: consonantCount, remainingSlots: remaining) {
            allowed.insert(.consonant)
        }

        return allowed
    }

    func isPickAllowed(
        _ kind: LetterKind,
        currentVowels: Int,
        currentConsonants: Int,
        remainingSlots: Int
    ) -> Bool {
        guard remainingSlots > 0 else { return false }

        let nextVowels = currentVowels + (kind == .vowel ? 1 : 0)
        let nextConsonants = currentConsonants + (kind == .consonant ? 1 : 0)
        return canStillSatisfyMinimums(vowels: nextVowels, consonants: nextConsonants, remainingSlots: remainingSlots - 1)
    }

    private func canStillSatisfyMinimums(vowels: Int, consonants: Int, remainingSlots: Int) -> Bool {
        let neededVowels = max(0, 1 - vowels)
        let neededConsonants = max(0, 1 - consonants)
        return neededVowels + neededConsonants <= remainingSlots
    }
}